File: code/golang/SocialMedia.go
package main
import "fmt"

const socialmediacount int = 35

type SocialMedia struct {
	title,
	url,
	desc,
	image,
	appd_id,
	redirect_url,
	via,
	hash_tags,
	provider,
	language,
	category,
	user_id,
	phone_number,
	email_address,
	cc_email_address,
	bcc_email_address,
	text string

	urls,
	urlsOrderedByPopularity [socialmediacount]string
}

func SocialMediaInterface() SocialMedia {
	sm := SocialMediaDefault()

	sm.urlsOrderedByPopularity = [...]string{
			"google.bookmarks",
			"facebook",
			"reddit",
			"whatsapp",
			"twitter",
			"linkedin",
			"tumblr",
			"pinterest",
			"blogger",
			"livejournal",
			"evernote",
			"add.this",
			"getpocket",
			"hacker.news",
			"buffer",
			"flipboard",
			"instapaper",
			"surfingbird.ru",
			"flattr",
			"diaspora",
			"qzone",
			"vk",
			"weibo",
			"ok.ru",
			"douban",
			"xing",
			"renren",
			"threema",
			"sms",
			"line.me",
			"skype",
			"telegram.me",
			"email",
			"gmail",
			"yahoo",
	}

	sm.text = sm.title

	if(len(sm.desc) > 0) {
		sm.text += "%20%3A%20"		// this is just: " : "
		sm.text += sm.desc
	}

	var addthis string = "http://www.addthis.com/bookmark.php?url=" + sm.url
	var blogger string = "https://www.blogger.com/blog-this.g?u=" + sm.url + "&n=" + sm.title + "&t=" + sm.desc
	var buffer string = "https://buffer.com/add?text=" + sm.text + "&url=" + sm.url
	var diaspora string = "https://share.diasporafoundation.org/?title=" + sm.title + "&url=" + sm.url
	var douban string = "http://www.douban.com/recommend/?url=" + sm.url + "&title=" + sm.title
	var email string = "mailto:" + sm.email_address + "?subject=" + sm.title + "&body=" + sm.desc
	var evernote string = "https://www.evernote.com/clip.action?url=" + sm.url + "&title=" + sm.text
	var getpocket string = "https://getpocket.com/edit?url=" + sm.url
	var facebook string = "http://www.facebook.com/sharer.php?u=" + sm.url
	var flattr string = "https://flattr.com/submit/auto?user_id=" + sm.user_id + "&url=" + sm.url + "&title=" + sm.title + "&description=" + sm.text + "&language=" + sm.language + "&tags=" + sm.hash_tags + "&hidden=HIDDEN&category=" + sm.category
	var flipboard string = "https://share.flipboard.com/bookmarklet/popout?v=2&title=" + sm.text + "&url=" + sm.url 
	var gmail string = "https://mail.google.com/mail/?view=cm&to=" + sm.email_address + "&su=" + sm.title + "&body=" + sm.url + "&bcc=" + sm.bcc_email_address + "&cc=" + sm.cc_email_address
	var googlebookmarks string = "https://www.google.com/bookmarks/mark?op=edit&bkmk=" + sm.url + "&title=" + sm.title + "&annotation=" + sm.text + "&labels=" + sm.hash_tags
	var instapaper string = "http://www.instapaper.com/edit?url=" + sm.url + "&title=" + sm.title + "&description=" + sm.desc
	var lineme string = "https://lineit.line.me/share/ui?url=" + sm.url + "&text=" + sm.text
	var linkedin string = "https://www.linkedin.com/sharing/share-offsite/?url=" + sm.url
	var livejournal string = "http://www.livejournal.com/update.bml?subject=" + sm.text + "&event=" + sm.url
	var hackernews string = "https://news.ycombinator.com/submitlink?u=" + sm.url + "&t=" + sm.title
	var okru string = "https://connect.ok.ru/dk?st.cmd=WidgetSharePreview&st.shareUrl=" + sm.url
	var pinterest string = "http://pinterest.com/pin/create/button/?url=" + sm.url
	var qzone string = "http://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?url=" + sm.url
	var reddit string = "https://reddit.com/submit?url=" + sm.url + "&title=" + sm.title
	var renren string = "http://widget.renren.com/dialog/share?resourceUrl=" + sm.url + "&srcUrl=" + sm.url + "&title=" + sm.text + "&description=" + sm.desc
	var skype string = "https://web.skype.com/share?url=" + sm.url + "&text=" + sm.text
	var sms string = "sms:" + sm.phone_number + "?body=" + sm.text
	var surfingbird string = "http://surfingbird.ru/share?url=" + sm.url + "&description=" + sm.desc + "&screenshot=" + sm.image + "&title=" + sm.title
	var telegramme string = "https://t.me/share/url?url=" + sm.url + "&text=" + sm.text + "&to=" + sm.phone_number
	var threema string = "threema://compose?text=" + sm.text + "&id=" + sm.user_id
	var tumblr string = "https://www.tumblr.com/widgets/share/tool?canonicalUrl=" + sm.url + "&title=" + sm.title + "&caption=" + sm.desc + "&tags=" + sm.hash_tags
	var twitter string = "https://twitter.com/intent/tweet?url=" + sm.url + "&text=" + sm.text + "&via=" + sm.via + "&hashtags=" + sm.hash_tags
	var vk string = "http://vk.com/share.php?url=" + sm.url + "&title=" + sm.title + "&comment=" + sm.desc
	var weibo string = "http://service.weibo.com/share/share.php?url=" + sm.url + "&appkey=&title=" + sm.title + "&pic=&ralateUid="
	var whatsapp string = "https://api.whatsapp.com/send?text=" + sm.text + "%20" + sm.url
	var xing string = "https://www.xing.com/spi/shares/new?url=" + sm.url
	var yahoo string = "http://compose.mail.yahoo.com/?to=" + sm.email_address + "&subject=" + sm.title + "&body=" + sm.text

		// NOTE : This ordering must identically match urlsOrderedByPopularity.

        sm.urls = [...]string{
		googlebookmarks,
		facebook,
		reddit,
		whatsapp,
		twitter,
		linkedin,
		tumblr,
		pinterest,
		blogger,
		livejournal,
		evernote,
		addthis,
		getpocket,
		hackernews,
		buffer,
		flipboard,
		instapaper,
		surfingbird,
		flattr,
		diaspora,
		qzone,
		vk,
		weibo,
		okru,
		douban,
		xing,
		renren,
		threema,
		sms,
		lineme,
		skype,
		telegramme,
		email,
		gmail,
		yahoo,
        }

	return sm
}

func SocialMediaDefault() SocialMedia {
	sm := SocialMedia{
		title:"EarthFluent",
		url:"http%3A%2F%2Fwww.earthfluent.com%2F",
		desc:"Want%20to%20learn%20another%20language%3F%20%20Learn%20Spanish%2C%20French%2C%20Italian%2C%20German%2C%20Japanese%2C%20Chinese%2C%20Hindi%2C%20Indonesian%2C%20Dutch%2C%20Polish%2C%20Portuguese%2C%20or%20Russian%21",
	}  
	return sm
}

func main(){
	sm := SocialMediaInterface()

	for i := 0; i < socialmediacount; i++ {
		fmt.Println(sm.urlsOrderedByPopularity[i] + " : " + sm.urls[i]);
	}
}

File: code/cpp/SocialMedia.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string.h>
using namespace std;

#define SOCIALMEDIASERVICES 37
#define URLSIZE 1024
#define ARGLENGTH 1024

class SocialMedia {
	public:
		SocialMedia();

		char url[ARGLENGTH] = "";
		char title[ARGLENGTH] = "";
		char desc[ARGLENGTH] = "";
		char image[ARGLENGTH] = "";
		char app_id[ARGLENGTH] = "";
		char redirect_url[ARGLENGTH] = "";
		char via[ARGLENGTH] = "";
		char hash_tags[ARGLENGTH] = "";
		char provider[ARGLENGTH] = "";
		char language[ARGLENGTH] = "";
		char user_id[ARGLENGTH] = "";
		char category[ARGLENGTH] = "";
		char phone_number[ARGLENGTH] = "";
		char email_address[ARGLENGTH] = "";
		char cc_email_address[ARGLENGTH] = "";
		char bcc_email_address[ARGLENGTH] = "";

		char urls[SOCIALMEDIASERVICES][URLSIZE];
		char urlsOrderedByPopularity[SOCIALMEDIASERVICES][URLSIZE] = {
			"google.bookmarks",
			"facebook",
			"reddit",
			"twitter",
			"linkedin",
			"tumblr",
			"pinterest",
			"blogger",
			"livejournal",
			"evernote",
			"add.this",
			"getpocket",
			"hacker.news",
			"digg",
			"buffer",
			"flipboard",
			"instapaper",
			"surfingbird.ru",
			"flattr",
			"diaspora",
			"qzone",
			"vk",
			"weibo",
			"ok.ru",
			"douban",
			"xing",
			"renren",
			"threema",
			"sms",
			"line.me",
			"skype",
			"telegram.me",
			"email",
			"gmail",
			"yahoo",
		};
		char urlsOrderedByName[SOCIALMEDIASERVICES][URLSIZE] = {};
};

SocialMedia::SocialMedia() {
	sprintf(url, "%s", "http%3A%2F%2Fwww.earthfluent.com%2F");
	sprintf(title, "%s", "EarthFluent");
	sprintf(desc, "%s", "Want%20to%20learn%20another%20language%3F%20%20Learn%20Spanish%2C%20French%2C%20Italian%2C%20German%2C%20Japanese%2C%20Chinese%2C%20Hindi%2C%20Indonesian%2C%20Dutch%2C%20Polish%2C%20Portuguese%2C%20or%20Russian%21");

	char text[ARGLENGTH];

	if(strlen(desc)) {
		sprintf(text, "%s%s%s", title, "%20%3A%20", desc);
	} else {
		sprintf(text, "%s", title);
	}

	std::vector<string> stringarray (urlsOrderedByPopularity, urlsOrderedByPopularity + SOCIALMEDIASERVICES);
	std::sort(stringarray.begin(), stringarray.end());

	for(int i = 0; i < SOCIALMEDIASERVICES; i++) {
		strcpy(urlsOrderedByName[i], stringarray[i].c_str());
	}

	char addthis[URLSIZE];
	sprintf(addthis, "%s%s", "http://www.addthis.com/bookmark.php?url=", url);
	
	char blogger[URLSIZE];
	sprintf(blogger, "%s%s%s%s%s%s", "https://www.blogger.com/blog-this.g?u=", url, "&n=", title, "&t=", desc);

	char buffer[URLSIZE];
	sprintf(buffer, "%s%s%s%s", "https://buffer.com/add?text=", text, "&url=", url);

	char diaspora[URLSIZE];
	sprintf(diaspora, "%s%s%s%s", "https://share.diasporafoundation.org/?title=", title, "&url=", url);
	
	char digg[URLSIZE];
	sprintf(digg, "%s%s%s%s", "http://digg.com/submit?url=", url, "&title=", text);

	char douban[URLSIZE];
	sprintf(douban, "%s%s%s%s", "http://www.douban.com/recommend/?url=", url, "&title=", title);

	char email[URLSIZE];
	sprintf(email, "%s%s%s%s%s%s", "mailto:", email_address, "?subject=", title, "&body=", desc);

	char evernote[URLSIZE];
	sprintf(evernote, "%s%s%s%s", "https://www.evernote.com/clip.action?url=", url, "&title=", title);
	
	char getpocket[URLSIZE];
	sprintf(getpocket, "%s%s", "https://getpocket.com/edit?url=", url);
	
	char facebook[URLSIZE];
	sprintf(facebook, "%s%s", "http://www.facebook.com/sharer.php?u=", url);
	
	char flattr[URLSIZE];
	sprintf(flattr, "%s%s%s%s%s%s%s%s%s%s%s%s%s%s", "https://flattr.com/submit/auto?user_id=", user_id, "&url=", url, "&title=", title, "&description=", text, "&language=", language, "&tags=", hash_tags, "&hidden=HIDDEN&category=", category);
	
	char flipboard[URLSIZE];
	sprintf(flipboard, "%s%s%s%s", "https://share.flipboard.com/bookmarklet/popout?v=2&title=", text, "&url=", url);
	
	char gmail[URLSIZE];
	sprintf(gmail, "%s%s%s%s%s%s%s%s%s%s", "https://mail.google.com/mail/?view=cm&to=", email_address, "&su=", title, "&body=", url, "&bcc=", bcc_email_address, "&cc=", cc_email_address);
	
	char googlebookmarks[URLSIZE];
	sprintf(googlebookmarks, "%s%s%s%s%s%s%s%s", "https://www.google.com/bookmarks/mark?op=edit&bkmk=", url, "&title=", title, "&annotation=", text, "&labels=", hash_tags);
	
	char instapaper[URLSIZE];
	sprintf(instapaper, "%s%s%s%s%s%s", "http://www.instapaper.com/edit?url=", url, "&title=", title, "&description=", desc);
	
	char lineme[URLSIZE];
	sprintf(lineme, "%s%s%s%s", "https://lineit.line.me/share/ui?url=", url, "&text=", text);
	
	char linkedin[URLSIZE];
	sprintf(linkedin, "%s%s", "https://www.linkedin.com/sharing/share-offsite/?url=", url);
	
	char livejournal[URLSIZE];
	sprintf(livejournal, "%s%s%s%s", "http://www.livejournal.com/update.bml?subject=", text, "&event=", url);
	
	char hackernews[URLSIZE];
	sprintf(hackernews, "%s%s%s%s", "https://news.ycombinator.com/submitlink?u=", url, "&t=", title);
	
	char okru[URLSIZE];
	sprintf(okru, "%s%s", "https://connect.ok.ru/dk?st.cmd=WidgetSharePreview&st.shareUrl=", url);
	
	char pinterest[URLSIZE];
	sprintf(pinterest, "%s%s", "http://pinterest.com/pin/create/button/?url=", url);
	
	char qzone[URLSIZE];
	sprintf(qzone, "%s%s", "http://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?url=", url);
	
	char reddit[URLSIZE];
	sprintf(reddit, "%s%s%s%s", "https://reddit.com/submit?url=", url, "&title=", title);
	
	char renren[URLSIZE];
	sprintf(renren, "%s%s%s%s%s%s%s%s", "http://widget.renren.com/dialog/share?resourceUrl=", url, "&srcUrl=", url, "&title=", title, "&description=", desc);
	
	char skype[URLSIZE];
	sprintf(skype, "%s%s%s%s", "https://web.skype.com/share?url=", url, "&text=", text);
	
	char sms[URLSIZE];
	sprintf(sms, "%s%s%s%s", "sms:", phone_number, "?body=", text);

	char surfingbird[URLSIZE];
	sprintf(surfingbird, "%s%s%s%s%s%s%s%s", "http://surfingbird.ru/share?url=", url, "&description=", desc, "&screenshot=", image, "&title=", title);
	
	char telegramme[URLSIZE];
	sprintf(telegramme, "%s%s%s%s%s%s", "https://t.me/share/url?url=", url, "&text=", text, "&to=", phone_number);
	
	char threema[URLSIZE];
	sprintf(threema, "%s%s%s%s", "threema://compose?text=", text, "&id=", user_id);
	
	char tumblr[URLSIZE];
	sprintf(tumblr, "%s%s%s%s%s%s%s%s", "https://www.tumblr.com/widgets/share/tool?canonicalUrl=", url, "&title=", title, "&caption=", desc, "&tags=", hash_tags);
	
	char twitter[URLSIZE];
	sprintf(twitter, "%s%s%s%s%s%s%s%s", "https://twitter.com/intent/tweet?url=", url, "&text=", text, "&via=", via, "&hashtags=", hash_tags);
	
	char vk[URLSIZE];
	sprintf(vk, "%s%s%s%s%s%s", "http://vk.com/share.php?url=", url, "&title=", title, "&comments=", desc);
	
	char weibo[URLSIZE];
	sprintf(weibo, "%s%s%s%s%s", "http://service.weibo.com/share/share.php?url=", url, "&appkey=&title=", title, "&pic=&ralateUid=");
	
	char xing[URLSIZE];
	sprintf(xing, "%s%s", "https://www.xing.com/spi/shares/new?url=", url);
	
	char yahoo[URLSIZE];
	sprintf(yahoo, "%s%s%s%s%s%s", "http://compose.mail.yahoo.com/?to=", email_address, "&subject=", title, "&body=", text);

	int i = 0;

		// NOTE : This ordering must identically match urlsOrderedByPopularity.
	
	strcpy(urls[i++], googlebookmarks);
	strcpy(urls[i++], facebook);
	strcpy(urls[i++], reddit);
	strcpy(urls[i++], twitter);
	strcpy(urls[i++], linkedin);
	strcpy(urls[i++], tumblr);
	strcpy(urls[i++], pinterest);
	strcpy(urls[i++], blogger);
	strcpy(urls[i++], livejournal);
	strcpy(urls[i++], evernote);
	strcpy(urls[i++], addthis);
	strcpy(urls[i++], getpocket);
	strcpy(urls[i++], hackernews);
	strcpy(urls[i++], digg);
	strcpy(urls[i++], buffer);
	strcpy(urls[i++], flipboard);
	strcpy(urls[i++], instapaper);
	strcpy(urls[i++], surfingbird);
	strcpy(urls[i++], flattr);
	strcpy(urls[i++], diaspora);
	strcpy(urls[i++], qzone);
	strcpy(urls[i++], vk);
	strcpy(urls[i++], weibo);
	strcpy(urls[i++], okru);
	strcpy(urls[i++], douban);
	strcpy(urls[i++], xing);
	strcpy(urls[i++], renren);
	strcpy(urls[i++], threema);
	strcpy(urls[i++], sms);
	strcpy(urls[i++], lineme);
	strcpy(urls[i++], skype);
	strcpy(urls[i++], telegramme);
	strcpy(urls[i++], email);
	strcpy(urls[i++], gmail);
	strcpy(urls[i++], yahoo);
};

int main() {
	SocialMedia sm;

	for(int i = 0; i < SOCIALMEDIASERVICES; i++) {
		printf("%s", sm.urlsOrderedByPopularity[i]);
		printf(" : ");
		printf("%s", sm.urls[i]);
		printf("\n");
	}

	return 0;
}
